[WIP] Bootstrap 4

alexfedoseev opened the pull request:
It's still WIP. Let's discuss [updated API and other initial stuff](https://github.com/shakacode/bootstrap-loader/tree/alex/bootstrap-4).

/cc: @justin808 @markusjwetzel

justin808 commented on .bootstraprc:
```diff
@@ -0,0 +1,85 @@
+---
```
I think this file should be bootstrap version specific.

maybe .bootstraprc-4.0

And we should allow a the specification of a custom one in case there's a new beta version of bootstrap or some other extensions.

alexfedoseev replied:
We should define and check only 3 or 4. Other should be in user land. It's gonna work as long as paths to js scss won't be changed. for major version of bootstrap.

justin808 commented on package.json:
```diff
@@ -1,14 +1,22 @@
   "version": "0.0.1",
-  "description": "Boostrap for Webpack",
-  "main": "index.js",
+  "description": "Boostrap 4 for Webpack",
```
Maybe just Bootstrap for Webpack if we support multiple versions?

alexfedoseev replied:
:+1:

justin808 commented on src/bootstrap.config.js:
```diff
@@ -0,0 +1,21 @@
+import fileExists from './utils/fileExists';
+
+
+const CONFIG_FILE = '.bootstraprc';
```
We should have different ones for each version of bootstrap.

alexfedoseev replied:
I'd rather put `bootstrapVersion: 4` in `.bootstraprc`, there is no use cases for multiple versions of this config in one project.

justin808 commented on src/bootstrap.config.js:
```diff
@@ -0,0 +1,21 @@
+);
+
+const configContent = stripComments(fs.readFileSync(configPath, 'utf8'));
+const config = yaml.safeLoad(configContent);
```
Yaml is better for editing :+1:

justin808 commented on src/bootstrap.loader.js:
```diff
@@ -0,0 +1,52 @@
+  if (this.cacheable) this.cacheable();
+
+  // Resolve `bootstrap` package
+  const bootstrapPath = resolveModule('bootstrap');
```
This might vary depending on the version of bootstrap.

alexfedoseev replied:
I'll handle it.

justin808 commented on src/bootstrap.scripts.loader.js:
```diff
@@ -0,0 +1,20 @@
+
+  return (
+    scriptsModules
+      .filter(module => config.scripts[module])
```
I was wondering if we need the filtering because if a minor release comes out with an additional file, you won't just be able to add it to your config.

alexfedoseev replied:
We filter only what user said to filter, it's his choice. He can easily define additional file in his config and it will be added to bundle.

justin808 replied:
I see!

justin808 commented on src/bootstrap.styles.loader.js:
```diff
@@ -0,0 +1,22 @@
+
+  return (
+    stylesModules
+      .filter(module => config.styles[module])
```
see comment above on filtering

justin808 commented on src/bootstrap.styles.loader.js:
```diff
@@ -0,0 +1,22 @@
+
+  const { bootstrapPath } = loaderUtils.parseQuery(this.query);
+
+  // const useFlexbox = config.flexbox;
```
?

alexfedoseev replied:
In Bootstrap 4 we must define do we want to use flexbox model or not. This configuration is WIP.

justin808 commented on src/utils/resolveModule.js:
```diff
@@ -0,0 +1,16 @@
+import resolve from 'resolve';
```
@alexfedoseev Looks like you found a better way to do the location of bootstrap.

justin808 commented:
first review done -- looks great

markusjwetzel commented:
I have nothing to add. I think you are on the right road! :+1:

alexfedoseev commented:
Looks like it's done, but it's not tested at all. Will finish this tomorrow.

alexfedoseev commented:
`bootstrap-loader@0.0.2` on npm. I updated [example app](https://github.com/shakacode/bootstrap-sass-loader-example/pull/6) to test it out.

justin808 commented on src/utils/joinLoaders.js:
```diff
@@ -0,0 +1,3 @@
+export default loaders => (
+  Array.isArray(loaders) && loaders.join('!') + '!'
```
semicolon? maybe not linted yet.

justin808 replied:
Maybe show the operator precedence?
```
(Array.isArray(loaders) && loaders.join('!')) + '!'
```

alexfedoseev replied:
Semicolon is on the next line.

Changed it to this:

```js
export default function(loaders) {
  if (!Array.isArray(loaders)) {
    throw new Error(`
      Specify your loaders as an array.
      Default is ['style', 'css', 'scss']
    `);
  }

  return loaders.map(loader => loader + '!').join('');
}
```

justin808 commented on src/utils/getFontsPath.js:
```diff
@@ -0,0 +1,5 @@
+import path from 'path';
+
+export default (bootstrapRelPath) => {
+  return path.join(bootstrapRelPath, 'assets', 'fonts', 'bootstrap/');
```
We need to be very careful for windows paths. There was a prior PR that dealt with a small change for windows.

alexfedoseev replied:
I thought `path` takes care of it, no? I haven't tested it on Windows though.

https://nodejs.org/api/path.html#path_path_normalize_p

justin808 commented on src/bootstrap.scripts.loader.js:
```diff
@@ -0,0 +1,16 @@
+module.exports = function() {
+  if (this.cacheable) this.cacheable();
+
+  const config = global.__BOOTSTRAP_CONFIG__;
```
why the global?

alexfedoseev replied:
It's because I need to pass normalized config to styles and scripts loader. If I pass it as stringified params, it slows down build: in common case when all bootstrap parts are enabled build time increases from ~2 sec to ~22 sec. As there is no way to pass non-serialized JS object to loader, I made it global.

justin808 commented on src/bootstrap.config.js:
```diff
@@ -0,0 +1,85 @@
+const CONFIG_FILE = '.bootstraprc';
+
+const userConfigPath = path.resolve(__dirname, `../../../${CONFIG_FILE}`);
+const isUserConfig = fileExists(userConfigPath);
```
How does the ../../.. ensure the right position?

alexfedoseev replied:
I assume `bootstrap-loader` will be installed in `node_modules`, which is located in root directory of project. And `.bootstraprc` will  be on the same level with `node_modules`.

justin808 commented on src/bootstrap.config.js:
```diff
@@ -0,0 +1,85 @@
+  }
+
+  const defaultConfigPath = (
+    path.resolve(__dirname, `../${CONFIG_FILE}-${bootstrapVersion}-default`)
```
function for this? with line 48?

alexfedoseev replied:
:+1:

justin808 commented on src/bootstrap.loader.js:
```diff
@@ -0,0 +1,98 @@
+
+  logger.debug(`Hey, we're in DEBUG mode! Yabba dabba doo!`);
+
+  logger.debug('Usung Bootstrap version:', bootstrapVersion);
```
Using

justin808 replied:
Maybe log.info?

alexfedoseev replied:
I think we shouldn't spam anything to console in default mode, if something went wrong — we will throw. And the real use case I can make up for it to be enabled is when user has faced some issue and we will ask him to turn debug mode on -> post output to help us find the problem. So debug is apt term here, what do you think?

justin808 commented on src/bootstrap.loader.js:
```diff
@@ -0,0 +1,98 @@
+  const bootstrapNPMModule = (
+    bootstrapVersion === 3 ? 'bootstrap-sass' : 'bootstrap'
+  );
+  logger.debug('Usung Bootstrap module:', bootstrapNPMModule);
```
Using

alexfedoseev replied:
Fixed.

justin808 commented on src/bootstrap.styles.loader.js:
```diff
@@ -0,0 +1,60 @@
+
+  if (bootstrapCustomizations) {
+    processedStyles.push(
+      createUserImport(bootstrapCustomizations, this)
```
Did you remove the mainSass?

Basically, we need at least 2 customizations (for Bootstrap 3), the pre-load and the post-load one.

The current loader supports 2 post loading ones:
1. A placeholder to customize bootstrap
2. Your own styles Sass manifest.

alexfedoseev replied:
It's not yet added, because I'm figuring out strategy for it.

Here is my conclusion.

There are 2 conventional use cases of loading CSS:

1. Common. When user provides one endpoint (SCSS file) where all CSS is collected via `import`s. This can be handled by `bootstrap-loader` as it uses same loaders for loading user's & bootstrap's css.
1. CSS Modules. There is no endpoint here, all CSS is loaded from JS components, so in this case `bootstrap-loader` will load Bootstrap and pass application endpoint to the next loaders (without touching them), where CSS Modules will be handled (on webpack config level, `config.module.loaders` section). We will provide example of it. So if user uses CSS Modules and will provide SCSS endpoint here (to some common CSS, which is not used as CSS Modules) — it will be loaded globally, just like Bootstrap styles.

alexfedoseev replied:
Added in b114b11. I renamed it to `appStyles`, b/c I figured out what `mainSass` is only after reading the code. Do you have any other options for name of this param?

justin808 commented on src/utils/createRequire.js:
```diff
@@ -0,0 +1,4 @@
+export default (module, disableProcessing) => {
+  const loader = disableProcessing ? '-!' + module : module;
```
for what reason do we pass disableProcessing as true?

alexfedoseev replied:
Yak shaving, voodoo coding. I was searching for what is causing increased build time (it's not it, see details above). Actually I took it from previous version of `bootstrap-sass-loader`. If it's for no reason there, I'll remove it from here.

justin808 commented on src/utils/createRequire.js:
```diff
@@ -0,0 +1,4 @@
+export default (module, disableProcessing) => {
+  const loader = disableProcessing ? '-!' + module : module;
+  return `require (${JSON.stringify(loader)});`;
```
What does the JSON.stringify do in this case? Just quote the string? If so, stringify required?

alexfedoseev replied:
It's used for user's input require also, so I decided to leave `JSON.stringify`.

justin808 commented on src/utils/parseConfig.js:
```diff
@@ -0,0 +1,8 @@
+import fs from 'fs';
+import yaml from 'js-yaml';
+import stripComments from 'strip-json-comments';
```
says strip-json

but we're using yaml

OK?

alexfedoseev replied:
`YAML` and `JSON`. `YAML` comments are stripped by parser, but `JSON` is passed as is. With this one users can have comments in their `JSON`ified `.bootstraprc`.

justin808 commented on src/utils/resolveModule.js:
```diff
@@ -0,0 +1,16 @@
+import resolve from 'resolve';
+
+export default module => {
```
Maybe show that module is a param to the fat arrow function?
```js
export default (module) => {
```

Also, it's not totally clear that the fat arrow is better than `function` when export default is used.


```js
export default function(module) {
```

alexfedoseev replied:
:+1:  for `export default function`

justin808 commented:
Looking awesome Alex!

alexfedoseev commented:
Pushed updates to this branch!

alexfedoseev commented:
Updated, needs review.

justin808 commented:
@alexfedoseev between times I review, can we squash the commits? then I can review just the commit after the prior, and after review, we squash?

alexfedoseev commented:
> @alexfedoseev between times I review, can we squash the commits? then I can review just the commit after the prior, and after review, we squash?

Sure!

alexfedoseev commented:
I updated paths to relatives to avoid cases, when user has `!` in his absolute path.
Coomit: ad312c1

And updated README.
Coomit: 91b6d3b
Link: [README.md](https://github.com/shakacode/bootstrap-loader/blob/alex/bootstrap-4/README.md)

Review needed.

justin808 commented:
@alexfedoseev how about we merge this and create issues for pending items and a list of what's needed for 1.0.

alexfedoseev commented:
Sure! Will do this tomorrow :+1:

mmahalwy commented:
any update on this guys? This is really exciting!

justin808 commented:
@mmahalwy: https://github.com/shakacode/react-webpack-rails-tutorial/issues/162

alexfedoseev commented:
@mmahalwy Actually you can try this out already, it's published to npm and here is the [draft of README with instructions](https://github.com/shakacode/bootstrap-loader/tree/alex/bootstrap-4) + [links to Examples](https://github.com/shakacode/bootstrap-loader/tree/alex/bootstrap-4#examples).

positron commented:
FYI the README has a broken link:

> Basic usage: https://github.com/shakacode/bootstrap-loader-example

justin808 commented:
https://github.com/shakacode/bootstrap-sass-loader-example/tree/alex/bootstrap-loader, @positron 

as well as https://github.com/shakacode/bootstrap-loader-css-modules-example

justin808 commented:
@alexfedoseev I didn't change this as it's on your branch.

alexfedoseev commented:
@justin808 Yep, we were talking about renaming the repo, I used updated link.

Maybe we should merge these repos to keep the stars/watchers, what do you think?
https://github.com/shakacode/bootstrap-sass-loader ->  https://github.com/shakacode/bootstrap-loader

https://github.com/shakacode/bootstrap-sass-loader-example ->  https://github.com/shakacode/bootstrap-loader-example

justin808 commented:
@alexfedoseev FABULOUS idea! I'll be around this weekend to help with this. Let's ROCK it!

alexfedoseev commented:
@justin808 I pushed latest changes + [updated README](https://github.com/shakacode/bootstrap-loader/blob/alex/bootstrap-4/README.md). But I faced another one issue when I use `resolve-url-loader` with `extract-text-webpack-plugin`.

I will merge this branch with master and create repo, reproducing the problem, to create an issue in `resolve-url-loader`.

alexfedoseev commented:
The problem is not with `extract-text-webpack-plugin` but with the fact, that we save our production build outside of the webpack's root directory (inside rails `app` folder), it causes issues with sourcemaps. Looking into it.

alexfedoseev commented:
I figured it out.
Issue with absolute / relative paths to entry point: https://github.com/shakacode/sass-rel-paths-issue
`sass-loader` or `resolve-url-loader` has source maps related issue with relative paths.

justin808 commented:
:+1: BTW, @sokra is great help on gitter for this stuff.

alexfedoseev commented:
Before crete an issue I tested it on case when we don't use `bootstrap-loader` to load styles and it worked. So problem is in how `bootstrap-loader` and `resolve-url-loader` works together with relative paths. I'll discuss it with author of `resolve-url-loader`.

alexfedoseev commented:
@justin808 It's fixed, please, review current master. The core is ready to ship, did I forget something here?

justin808 commented on .babelrc:
```diff
@@ -0,0 +1,3 @@
+{
+  "presets": ["es2015"]
```
going Babel 6?

alexfedoseev replied:
Yep.

justin808 commented on README.md:
````diff
@@ -1,3 +1,280 @@
+Get it via npm:
+
+```bash
+npm install bootstrap-loader
````
Keep same github repo, but new loader name. Seems OK. We should push another version that does some sort of install message or maybe a console message, or maybe just an update in the README.md.

alexfedoseev replied:
Will do after `1.0.0`.

justin808 commented on README.md:
````diff
@@ -1,3 +1,280 @@
+  button: true
+```
+
+```json
````
```
  ```js
```

alexfedoseev replied:
Done.

justin808 commented on README.md:
````diff
@@ -1,3 +1,280 @@
+entry: [ 'bootstrap-loader', './app' ]
+```
+
+Config is optional. It can be placed in root dir with name `.bootstraprc`. You can write it in `YAML` or `JSON` formats.
````
See this file <BLAH> for the default config file. Note, we don't recommend this as upgrades might make changes that will break your app.

alexfedoseev replied:
Updated in another PR.

justin808 commented on README.md:
```diff
@@ -1,3 +1,280 @@
+
+Default: `false`
+
+Extract styles to stand-alone css file using `extract-text-webpack-plugin`.
```
See [extract-text-plugin](https://github.com/webpack/extract-text-webpack-plugin) for more details.

alexfedoseev replied:
Updated in another PR.

justin808 commented on README.md:
```diff
@@ -1,3 +1,280 @@
+
+Check out example apps:
+
+* Basic usage: https://github.com/shakacode/bootstrap-loader-example
```
https://github.com/shakacode/bootstrap-sass-loader-example

or we create one for that URL you specify?

alexfedoseev replied:
I moved all examples in the sub folder of loader's repo.

justin808 commented on README.md:
```diff
@@ -1,3 +1,280 @@
+Check out example apps:
+
+* Basic usage: https://github.com/shakacode/bootstrap-loader-example
+* With CSS Modules: https://github.com/shakacode/bootstrap-loader-css-modules-example
```
This example shows off hot reloading as well!

alexfedoseev replied:
Updated in another PR.

justin808 commented on README.md:
```diff
@@ -1,3 +1,280 @@
+
+Default: `disabled`
+
+Outputs debugging info. In case you have troubles with loader you can set this option to `debug` and send us output form console.
```
Set this option to `debug` to output debugging information. This is critical for debugging issues. The output will go to your webpack console.

justin808 replied:
I'm not sure I like `disabled` for the default log setting.

logLevel is usually debug, info, warn, error.

So I'd say the default is `warn`. We certainly should log in some cases!

alexfedoseev replied:
Actually this option is commented out by default, and in case if something is wrong — we are throwing. There is nothing to warn about, so if processing went smooth — output is clear. If we found some mistakes — they are thrown.

justin808 commented on README.md:
```diff
@@ -1,3 +1,280 @@
+
+Here are options for Bootstrap 3 & 4.
+
+### Bootstrap 3
```
### Common Options for Bootstrap 3 and 4

alexfedoseev replied:
Updated in another PR.

justin808 commented on README.md:
```diff
@@ -1,3 +1,280 @@
+
+Default: `[ 'style', 'css', 'sass' ]`
+
+Array of webpack loaders. `sass-loader` is required, order matters.
```
The `style` loader should definitely go first and the `sass` loader should be last.

alexfedoseev replied:
Updated in another PR.

justin808 commented on README.md:
```diff
@@ -1,3 +1,280 @@
+
+Default: `disabled`
+
+This gets loaded after bootstrap variables is loaded. Thus, you may customize Bootstrap variables based on the values established in the Bootstrap [`_variables.scss`](https://github.com/twbs/bootstrap-sass/blob/master/assets/stylesheets/bootstrap/_variables.scss) file.
```
Note, if `bootstrap` did not have derived values, it would not be necessary to have two config files for customizing bootstrap values. 

If you want your bootstrap override value to apply to derived variable values, then place your customizations in `preBootstrapCustomizations`. If you want to be sure your changes don't affect other derived values, place the changes in `bootstrapCustomizations`.

If you are not sure, you can probably simply use `preBootstrapCustomizations` and, if you have issues, see  [`_variables.scss`](https://github.com/twbs/bootstrap-sass/blob/master/assets/stylesheets/bootstrap/_variables.scss) for derived values.

alexfedoseev replied:
Updated in another PR.

justin808 commented on README.md:
```diff
@@ -1,3 +1,280 @@
+
+If you use `bootstrap-loader` to load your styles (via `preBootstrapCustomizations`, `bootstrapCustomizations` & `appStyles`) and you load custom assets (fonts, images etc.), then you can use relative paths inside `url` method (relative to SASS file, from which you load asset).
+
+This was made possible thanks to [resolve-url-loader](https://github.com/bholloway/resolve-url-loader). In common case you don't have to do anything special to apply it — we are doing it internally (just don't forget to install it). But if you want to use its custom settings, please provide it explicitly via `styleLoaders` option in `.bootstraprc`:
```
We need an example of this.

Is the relative URL to the sass file? to the generated sass file during deployment? This is not clear.

alexfedoseev replied:
> We need an example of this.

I'll add it in another PR.

> Is the relative URL to the sass file?

Yep, I mentioned it in previous paragraph.

alexfedoseev replied:
Added in https://github.com/shakacode/bootstrap-loader/commit/3d4f36248dd8ca92f7b491fd254b52250635baa5.
